dropwizard-auth/src/main/java/io/dropwizard/auth/CachingAuthorizer.java

```diff
@@ -0,0 +1,172 @@
+package io.dropwizard.auth;
+
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import com.google.common.base.Predicate;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheBuilderSpec;
+import com.google.common.cache.CacheStats;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import java.security.Principal;
+
+import static com.codahale.metrics.MetricRegistry.name;
+
+/**
+ * An {@link Authorizer} decorator which uses a Guava {@link Cache} to
+ * temporarily cache principals' role associations.
+ *
+ * Cache entries include both inclusion and exclusion of a principal
+ * within a given role.
+ *
+ * @param <P> the type of principals on which the authorizer operates
+ */
+public class CachingAuthorizer<P extends Principal> implements Authorizer<P> {
+    private final Authorizer<P> underlying;
+    private final Cache<ImmutablePair<P, String>, Boolean> cache;
```

Comment:
I'm all for improving clarity, but because Java enum types are defined as classes, references to them can also be `null`. For example, the following is valid code:

    enum Foo { Allow, Deny }    
    final Foo foo = null;

If it seems ambiguous to use boolean values within the context of authorization (i.e. if it's not obvious that true=authorized, false=unauthorized), then an `Allow`/`Deny` enum would be more explicit. However, given that all of this is private code internal to the class, I think it's a fair tradeoff to use booleans to avoid introducing a new class. I think an inline comment on the cache type semantics is definitely warranted.